core/src/main/java/kafka/server/ClientRequestQuotaManager.java

```diff
@@ -0,0 +1,110 @@
+ */
+package kafka.server;
+
+import kafka.network.RequestChannel;
+
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.QuotaViolationException;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.server.config.ClientQuotaManagerConfig;
+import org.apache.kafka.server.quota.ClientQuotaCallback;
+import org.apache.kafka.server.quota.QuotaType;
+import org.apache.kafka.server.quota.QuotaUtils;
+
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import scala.jdk.javaapi.CollectionConverters;
+import scala.jdk.javaapi.OptionConverters;
+import scala.runtime.BoxedUnit;
+
+@SuppressWarnings("this-escape")
+public class ClientRequestQuotaManager extends ClientQuotaManager {
+    // Since exemptSensor is for all clients and has a constant name, we do not expire exemptSensor and only
+    // create once.
+    static final double NANOS_TO_PERCENTAGE_PER_SECOND = 100.0 / TimeUnit.SECONDS.toNanos(1);
+    private static final long DEFAULT_INACTIVE_EXEMPT_SENSOR_EXPIRATION_TIME_SECONDS = Long.MAX_VALUE;
+    private static final String EXEMPT_SENSOR_NAME = "exempt-" + QuotaType.REQUEST;
+
+    private final long maxThrottleTimeMs;
+    private final Metrics metrics;
+    private final MetricName exemptMetricName;
+    // Visible for testing
+    private final Sensor exemptSensor;
+
+    public ClientRequestQuotaManager(ClientQuotaManagerConfig config, Metrics metrics, Time time, String threadNamePrefix, Optional<ClientQuotaCallback> quotaCallback) {
+        super(config, metrics, QuotaType.REQUEST, time, threadNamePrefix, OptionConverters.toScala(quotaCallback));
+        this.maxThrottleTimeMs = TimeUnit.SECONDS.toMillis(config.quotaWindowSizeSeconds);
+        this.metrics = metrics;
+        this.exemptMetricName = metrics.metricName("exempt-request-time", QuotaType.REQUEST.toString(), "Tracking exempt-request-time utilization percentage");
+        exemptSensor = getOrCreateSensor(EXEMPT_SENSOR_NAME, DEFAULT_INACTIVE_EXEMPT_SENSOR_EXPIRATION_TIME_SECONDS, sensor -> sensor.add(exemptMetricName, new Rate()));
+    }
+
+    public Sensor exemptSensor() {
+        return exemptSensor;
+    }
+
+    private void recordExempt(double value) {
+        exemptSensor.record(value);
+    }
+
+    /**
+     * Records that a user/clientId changed request processing time being throttled. If quota has been violated, return
+     * throttle time in milliseconds. Throttle time calculation may be overridden by sub-classes.
+     * @param request client request
+     * @return Number of milliseconds to throttle in case of quota violation. Zero otherwise
+     */
+    public int maybeRecordAndGetThrottleTimeMs(RequestChannel.Request request, long timeMs) {
+        if (quotasEnabled()) {
+            request.setRecordNetworkThreadTimeCallback(timeNanos -> {
```

Comment:
Could you please change the input type of `setRecordNetworkThreadTimeCallback` from `Long => Unit` to `java.util.function.Consumer`? Using the scala type in java is a bit weird